content/blog/2020-12-20-bril-algebraic-data-types/index.md

```diff
@@ -0,0 +1,538 @@
+
+The type checker was also not terribly difficult, although it was a
+bit tedious because I had to add rules for every Bril
+instruction. Figuring out how to type-check `match` instructions
+within the dataflow analysis framework was a bit challenging.
+
+The Rust compiler was by far the most difficult and time-consuming
+part of the project. First, I spent a fair amount of time trying to
+figure out how to parse Rust source code before settling on the `syn`
+package. Figuring out how to use the parsed AST was also an involved
+process because it's a full AST for a "real" language, not just a toy
+language like Bril. I also spent quite a bit of time figuring out how
+to recursively tile expressions and how to keep track of variant
+types. It's also worth mentioning that I'm still a little new to Rust,
+so I was occasionally roadblocked by things like borrowing issues,
+although I am definitely getting better.
+
+
+## Evaluation
+
+### Correctness testing
+
+I created three sample Bril programs using algebraic types in the
+[`algebraic-types/samples`](https://github.com/Calsign/bril/tree/algebraic-types/algebraic-types/samples)
+directory as well as two sample Rust programs in the
+[`algebraic-types/samples-rs`](https://github.com/Calsign/bril/tree/algebraic-types/algebraic-types/samples-rs)
+directory.
+
+I tested the correctness of the `brili` additions by running the
+sample Bril programs and observing the output. The samples provide
+complete coverage of the added instructions, which isn't terribly
+difficult because there are only four instructions.
+
+I tested the correctness of the type checker by running it on some of
+the provided Bril benchmarks as well as the Bril samples for testing
+algebraic types. I also ran it on modified versions of these programs
+that were designed to fail the type checker to verify that it failed
+accordingly. I probably have not achieved complete coverage with
+respect to the core Bril instructions, but I have tested extensively
+with the algebraic type instructions and definitely have achieved
+complete coverage there.
+
+I tested the Rust compiler in several ways. I only tested it with two
+sample programs, but the two programs together cover all of the
+supported Rust syntax. I converted the output to the textual Bril
+format and examined the code to verify its correctess; I then ran it
+through the type checker to verify that the output programs have valid
+types; and finally, I ran it through `brili` to ascertain that the
+compiled programs output the expected values.
+
+### Reflection
+
+I think that using a powerful `match` instruction was the right
+decision. Having the ability to perform type-checking made me a lot
+more confident in the correctness of the Rust compiler.
+
+I do think that it would have been preferable to make the `construct`
+instruction require explicitly specifying the index of the constructor
+to construct. This approach would have been better for a number of
```

Comment:
Yeah, probably so.

Reply:
So if I revise the `construct` instruction to work this way, would that be good to go for a merge?

Also, is there any interest in the Rust compiler in the main repo? (I know @evanbenadler's proposed a more comprehensive implementation - perhaps that one would be better?)

Reply:
I think so. I'm still a little puzzled by what to do about the corresponding destructor too… a "safe" version is nice indeed, but the fact that the instruction lies about its type is troubling. I think it might make sense to merge it anyway and think about a different representation later.

Reply:
About the Rust-to-Bril compiler: why not?! Seems like fun.